File: Array/Count_DistinctEles_everyWindow.py
def CountDistinct(arr, n, k):
    res = []
    maxN = max(arr)
    hashT = [0]*(maxN+1)

    dis_count = 0
    for i in range(0, k):
        if hashT[arr[i]] == 0:
            dis_count = dis_count+1
        
        hashT[arr[i]] = hashT[arr[i]]+1
    
    res.append(dis_count)

    for i in range(k, n):
        if hashT[arr[i-k]] == 1:
            dis_count = dis_count-1
        
        hashT[arr[i-k]] = hashT[arr[i-k]]-1
        
        if hashT[arr[i]] == 0:
            dis_count = dis_count+1
        
        hashT[arr[i]] = hashT[arr[i]]+1
        
        res.append(dis_count)
    
    return res




if __name__ == "__main__":
    print("\n")
    # arr = [1,2,1,3,4,2,3]
    # k = 4
    # arr = [4,1,1]
    # k = 2
    n, k = map(int, input("\nEnter the Size of the Array and K window: ").split())
    arr = list(map(int, input("Enter the Array Elements: ").split()))[:n]

    ans = CountDistinct(arr, n, k)
    print("\nNumber of Distinct Elements in Every Window are:", ans)

    print("\n")

File: Cpp Codes/Count_DistinctEles_everyWindow.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

int CountElements(int *arr, int n, int k)
{
    int i;
    int maxN = arr[0];
    for(i=0; i<n; i++)
        maxN = max(maxN, arr[i]);
    
    vector<int> hashT(maxN+1, 0);

    int count = 0;
    vector<int> res;
    for(i=0; i<k; i++)
    {
        if(hashT[arr[i]] == 0)
            count = count+1;
        
        hashT[arr[i]] += 1;
    }
    res.push_back(count);

    for(i=k; i<n; i++)
    {
        if(hashT[arr[i-k]] == 1)
            count = count-1;
        
        hashT[arr[i-k]] -= 1;

        if(hashT[arr[i]] == 0)
            count = count+1;
        
        hashT[arr[i]] += 1;
        res.push_back(count);
    }

    for(auto it=res.begin(); it != res.end(); it++)
        cout<<*it<<" ";
}

int main()
{
    // int arr[] = {1,2,1,3,4,2,3};
    // int n = sizeof(arr)/sizeof(arr[0]);
    // int k = 4;
    int n, i, k;
    cout<<"\nEnter the size and k: ";
    cin>>n>>k;

    int arr[n];
    cout<<"Enter the Array Elements: ";
    for(i=0; i<n; i++)
        cin>>arr[i];
    cout<<"\n";

    CountElements(arr, n, k);
    
    cout<<"\n";
    return 0;
}